File: sword-means-offer-2/003/main.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
# @Time     : 2021/12/12 19:55
# @Author   : Ranshi
# @File     : 123.py
# @Doc      : 剑指 Offer II 003. 前 n 个数字二进制中 1 的个数
class Solution:
    def countBits(self, n: int) -> list[int]:
        bits: list[int] = [0] * (n + 1)
        for i in range(1, n + 1):
            bits[i] = bits[i & (i - 1)] + 1
        return bits


if __name__ == "__main__":
    print(Solution().countBits(n=2))

File: sword-means-offer-2/003/cpp/main.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
  int countOnes(int x)
  {
    int ones = 0;
    while (x > 0) {
      x &= (x - 1);
      ones++;
    }
    return ones;
  }
  vector<int> countBits(int n)
  {
    vector<int> bits(n + 1);
    for (size_t i = 0; i <= n; i++) {
      bits[i] = countOnes(i);
    }
    return bits;
  }
};

int
main(int argc, char const* argv[])
{
  auto v = Solution().countBits(2);
  for (size_t i = 0; i < v.size(); i++) {
    cout << v.at(i);
  }

  return 0;
}
